src/main/java/net/jonathangiles/tools/teenyhttpd/implementation/ReflectionUtils.java

```diff
@@ -0,0 +1,156 @@
+        }
+
+        if (actualTypeArgument instanceof ParameterizedType) {
+
+            ParameterizedType parameterizedType = (ParameterizedType) actualTypeArgument;
+
+            return new ParameterType((Class<?>) parameterizedType.getActualTypeArguments()[0],
+                    (Class<?>) parameterizedType.getRawType());
+        }
+
+        throw new IllegalStateException("Unknown type: " + actualTypeArgument);
+    }
+
+    public static Map<String, Field> getFields(Class<?> clazz) {
+
+        Map<String, Method> methodMap = Arrays.stream(clazz.getDeclaredMethods())
+                .collect(Collectors.toMap(Method::getName, m -> m));
+
+        return Arrays.stream(clazz.getDeclaredFields())
+                .filter(ReflectionUtils::isWritable)
+                .collect(Collectors.toMap(field -> getFieldName(field, methodMap), f -> f));
+    }
+
+    private static String getFieldName(Field field, Map<String, Method> methodMap) {
+
+        Method accessorMethod;
+
+        String baseName = field.getName().substring(0, 1).toUpperCase() + field.getName().substring(1);
+
+        if (field.getType() == boolean.class) {
+            accessorMethod = methodMap.get("is" + baseName);
+        } else {
+            accessorMethod = methodMap.get("get" + baseName);
+        }
+
+        if (accessorMethod != null) {
+            JsonAlias jsonAlias = accessorMethod.getAnnotation(JsonAlias.class);
+
+            if (jsonAlias != null) {
+                return jsonAlias.value();
+            }
+        }
+
+        return field.getName();
+    }
+
+    public static Map<String, Method> getMutators(Class<?> clazz) {
+        Map<String, Method> methodMap = Arrays.stream(clazz.getDeclaredMethods())
+                .collect(Collectors.toMap(Method::getName, m -> m));
+
+        Map<String, Method> resultMap = new HashMap<>();
+
+        for (Method method : methodMap.values()) {
+            if (!isMutator(method)) continue;
+
+            String mutatorName = getMutatorName(method, methodMap);
+
+            resultMap.put(mutatorName, method);
+        }
+
+        return resultMap;
+    }
+
+    /**
+     * Here we seek for the alias key in the accessor method, if it exists.
+     */
+    private static String getMutatorName(Method method, Map<String, Method> methodMap) {
+
+        String baseName = method.getName().substring(3);
+
+        Method accessorMethod;
+
+        if (method.getParameterTypes()[0] == boolean.class) {
+            accessorMethod = methodMap.get("is" + baseName);
+        } else {
+            accessorMethod = methodMap.get("get" + baseName);
+        }
+
+        if (accessorMethod != null) {
+            JsonAlias jsonAlias = accessorMethod.getAnnotation(JsonAlias.class);
+
+            if (jsonAlias != null) {
+                return jsonAlias.value();
+            }
+        }
+
+        return method.getName().substring(3, 4).toLowerCase()
+                + method.getName().substring(4);
```

Comment:
This is largely equivalent to the `getFieldName` method above. I wonder if we could reduce the repetition?

Reply:
I moved the repeated code to:

```java
private static String findAlias(Map<String, Method> methodMap, String baseName, Class<?> type) {
        Method accessorMethod;

        if (type == boolean.class) {
            accessorMethod = methodMap.get("is" + baseName);
        } else {
            accessorMethod = methodMap.get("get" + baseName);
        }

        if (accessorMethod != null) {
            JsonAlias jsonAlias = accessorMethod.getAnnotation(JsonAlias.class);

            if (jsonAlias != null) {
                return jsonAlias.value();
            }
        }

        return null;
    }
```